model/roast.go

```diff
@@ -0,0 +1,221 @@
+// Package model roast implementation.
+package model
+
+import (
+	"log"
+	"math"
+	"strings"
+	"time"
+
+	"github.com/satori/go.uuid"
+)
+
+// RoastMessage represents a general Roast message.
+type RoastMessage struct {
+	Hash        uuid.UUID `json:"hash"`
+	Row         uint      `json:"row"`
+	Column      uint      `json:"column"`
+	Engine      string    `json:"engine"`
+	Name        string    `json:"name"`
+	Description string    `json:"description"`
+}
+
+// RoastError represents a Roast error message.
+type RoastError struct {
+	RoastMessage
+}
+
+// RoastWarning represents a Roast error message.
+type RoastWarning struct {
+	RoastMessage
+}
+
+// RoastResult represent a Roast result.
+type RoastResult struct {
+	Username   string         `json:"username"`
+	Code       string         `json:"code"`
+	Score      uint           `json:"score"`
+	Language   string         `json:"language"`
+	Errors     []RoastError   `json:"errors"`
+	Warnings   []RoastWarning `json:"warnings"`
+	CreateTime time.Time      `json:"create_time"`
+}
+
+// AddError adds an error to the RoastResult.
+func (r *RoastResult) AddError(hash uuid.UUID, row, column uint, engine, name, description string) {
+	r.Errors = append(r.Errors, RoastError{
+		RoastMessage{
+			Hash:        hash,
+			Row:         row,
+			Column:      column,
+			Engine:      engine,
+			Name:        name,
+			Description: description,
+		},
+	})
+}
+
+// AddWarning adds an warning to the RoastResult.
+func (r *RoastResult) AddWarning(hash uuid.UUID, row, column uint, engine, name, description string) {
+	r.Warnings = append(r.Warnings, RoastWarning{
+		RoastMessage{
+			Hash:        hash,
+			Row:         row,
+			Column:      column,
+			Engine:      engine,
+			Name:        name,
+			Description: description,
+		},
+	})
+}
+
+// sloc implements a naĩve line counter for code.
+// All newlines are counted, so even empty rows and comments are counted.
+func (r *RoastResult) sloc() int {
```

Comment:
Note: Yes this will toast the CPU and fill the RAM for __large__ code snippets - it _should_ have been implemented using streams and stuff.

Reply:
Good that you comment this though. For when we write the report.